File: auto_translate.js
import fs from 'fs';
import path from 'path';
import { translate } from '@vitalets/google-translate-api';

const localesDir = './src/locales/locales';
const enPath = path.join(localesDir, 'en.json');

const enData = JSON.parse(fs.readFileSync(enPath, 'utf8'));

// Helper to delay
const sleep = ms => new Promise(r => setTimeout(r, ms));

async function translateObject(targetObj, enObj, langCode, prefix = '') {
  let modified = false;
  
  for (const key in enObj) {
    if (typeof enObj[key] === 'object' && enObj[key] !== null) {
      if (!targetObj[key]) targetObj[key] = {};
      const childModified = await translateObject(targetObj[key], enObj[key], langCode, prefix + key + '.');
      if (childModified) modified = true;
    } else if (typeof enObj[key] === 'string') {
      // Check if missing or identical
      if (!targetObj[key] || targetObj[key] === enObj[key]) {
        // Skip translating numbers or extremely short words like "2026", "DMCA" if they are identical, though it's safer to translate all
        if (enObj[key].length <= 2 && !/[a-zA-Z]/.test(enObj[key])) {
            targetObj[key] = enObj[key];
            continue;
        }

        try {
          console.log(`Translating [${langCode}] ${prefix}${key}: "${enObj[key].substring(0, 30)}..."`);
          const res = await translate(enObj[key], { to: langCode });
          targetObj[key] = res.text;
          modified = true;
          await sleep(200); // Prevent rate limiting
        } catch (error) {
          console.error(`Error translating [${langCode}] ${prefix}${key}:`, error.message);
          // Stop if rate limited to avoid spam
          if (error.message.includes('TooManyRequests')) {
              throw error;
          }
        }
      }
    }
  }
  return modified;
}

async function main() {
  const files = fs.readdirSync(localesDir).filter(f => f.endsWith('.json') && f !== 'en.json');
  
  for (const file of files) {
    const langCode = file.replace('.json', '');
    const filePath = path.join(localesDir, file);
    
    // Map some language codes if needed for google translate
    let gLang = langCode;
    if (gLang === 'zh-CN') gLang = 'zh-cn';
    if (gLang === 'zh-TW') gLang = 'zh-tw';
    if (gLang === 'pt-BR') gLang = 'pt';
    
    console.log(`\n--- Processing ${file} ---`);
    let targetData = {};
    try {
      targetData = JSON.parse(fs.readFileSync(filePath, 'utf8'));
    } catch(e) {
      targetData = {};
    }
    
    const modified = await translateObject(targetData, enData, gLang);
    
    if (modified) {
      fs.writeFileSync(filePath, JSON.stringify(targetData, null, 2));
      console.log(`Saved ${file}`);
    } else {
      console.log(`No changes for ${file}`);
    }
  }
}

main().catch(console.error);

File: auto_translate.py
import json
import os
import time
from deep_translator import GoogleTranslator

locales_dir = r"c:\Users\newFUTURE\Desktop\xmax2\Savesnapfast\src\locales\locales"
en_path = os.path.join(locales_dir, "en.json")

with open(en_path, 'r', encoding='utf-8') as f:
    en_data = json.load(f)

def translate_obj(target_obj, en_obj, lang_code, prefix=""):
    modified = False
    
    for key, en_val in en_obj.items():
        if isinstance(en_val, dict):
            if key not in target_obj:
                target_obj[key] = {}
            if translate_obj(target_obj[key], en_val, lang_code, f"{prefix}{key}."):
                modified = True
        elif isinstance(en_val, str):
            target_val = target_obj.get(key)
            
            if target_val is None or target_val == en_val:
                # skip short words
                if len(en_val) <= 2 and not any(c.isalpha() for c in en_val):
                    target_obj[key] = en_val
                    continue
                    
                print(f"Translating [{lang_code}] {prefix}{key}: {en_val[:30]}...")
                try:
                    translated = GoogleTranslator(source='en', target=lang_code).translate(en_val)
                    target_obj[key] = translated
                    modified = True
                    time.sleep(0.2) # sleep to avoid rate limits
                except Exception as e:
                    print(f"Error translating: {e}")
                    # Sleep longer on error
                    time.sleep(5.0)
                    
    return modified

def get_google_lang(lang_code):
    mapping = {
        'zh-CN': 'zh-CN',
        'zh-TW': 'zh-TW',
        'in': 'id', # Indonesian
        'pt-BR': 'pt'
    }
    return mapping.get(lang_code, lang_code.split('-')[0])

for filename in os.listdir(locales_dir):
    if not filename.endswith(".json") or filename == "en.json":
        continue
        
    lang_code = filename.replace(".json", "")
    g_lang = get_google_lang(lang_code)
    
    file_path = os.path.join(locales_dir, filename)
    print(f"\n--- Processing {filename} ---")
    
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            target_data = json.load(f)
    except Exception:
        target_data = {}
        
    modified = translate_obj(target_data, en_data, g_lang)
    
    if modified:
        with open(file_path, 'w', encoding='utf-8') as f:
            json.dump(target_data, f, ensure_ascii=False, indent=2)
        print(f"Saved {filename}")
    else:
        print(f"No changes for {filename}")
